### PR TITLE
stash: Prevent ID reuse for compute instances

### LINKED ISSUE #13100: Coordinator panics when `TAIL`ing a dropped cluster
This looks just like #12951, except that @heeringa observed it on 5ae377002bc1aa4a9ab6a037f6fc0bc8e9f034fe, which seemingly includes the fix for #12951 (https://github.com/MaterializeInc/materialize/pull/12978).

```
2022-06-13T21:31:37.666919Z  WARN mz_dataflow_types::client::replicated: Rehydrating replica 3
2022-06-13T21:31:37.667115Z  WARN mz_dataflow_types::client::replicated: Rehydrating replica 2
2022-06-13T21:32:07.193067Z  INFO mz_coord::catalog: drop table materialize.public.ten_k (u16)
2022-06-13T21:32:10.744437Z  INFO mz_coord::catalog: create table materialize.public.ten_k (u27)
2022-06-13T21:32:43.465334Z  INFO mz_coord::catalog: create view materialize.public.fifty_m (u28)
thread 'coordinator' panicked at 'called `Result::unwrap()` on an `Err` value: IdentifierMissing(User(25))', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:4958:47
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::result::unwrap_failed
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1785:5
   3: core::result::Result<T,E>::unwrap
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1078:23
   4: mz_coord::coord::Coordinator<S>::drop_sinks::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:4958:17
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
   6: mz_coord::coord::Coordinator<S>::clear_transaction::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:1824:36
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
   8: mz_coord::coord::Coordinator<S>::sequence_end_transaction_inner::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:3286:50
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  10: mz_coord::coord::Coordinator<S>::sequence_end_transaction::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:3252:66
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  12: mz_coord::coord::Coordinator<S>::message_command::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:1364:67
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  14: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.34/src/instrument.rs:272:9
  15: mz_coord::coord::Coordinator<S>::message_command::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:1203:5
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  17: mz_coord::coord::Coordinator<S>::serve::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:865:67
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  19: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/park/thread.rs:263:54
  20: tokio::coop::with_budget::{{closure}}
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:102:9
  21: std::thread::local::LocalKey<T>::try_with
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:442:16
  22: std::thread::local::LocalKey<T>::with
             at ./rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:418:9
  23: tokio::coop::with_budget
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:95:5
  24: tokio::coop::budget
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/coop.rs:72:5
  25: tokio::park::thread::CachedParkThread::block_on
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/park/thread.rs:263:31
  26: tokio::runtime::enter::Enter::block_on
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/enter.rs:151:13
  27: tokio::runtime::handle::Handle::block_on
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.19.1/src/runtime/handle.rs:296:9
  28: mz_coord::coord::serve::{{closure}}::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03966d7357c9ad138-1/materialize/tests/src/coord/src/coord.rs:5319:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Stream closed EOF for environment-001bd74d-48fe-4be0-998e-cb3eca622a5d-0/coordd-0 (coordd)
```

### PR DESCRIPTION
**Pull request by jkosh44**
The default behavior for ID selection in stash collections is to take
the maximum of all current IDs and add one. This means that after
deleting an item in a stash collection, it's ID can be reused, if the
collection is using the default behavior.

This is problematic for compute instances due to how TAIL is
implemented. TAIL creates a temporary sink that stores the ID of the
compute instance that it was created in. When the TAIL is canceled,
the associated sink is deleted from the compute instance, if the
compute instance still exists. The compute instance ID is used to
identify if the compute instance exists. The TAIL can outlive the
compute instance if the compute instance is deleted while the TAIL is
running. If the compute instance is deleted and new one is created,
then the new compute instance may share the same ID as the old one. The
TAIL will then mistakenly try and delete the sink from this new compute
instance and cause a panic.

This commit changes ID selection for compute instances so that they are
always monotonically increasing and no ID is ever reused. The
implementation is based on GlobalId allocation.

Fixes #13100

### Motivation
This PR fixes a recognized bug.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/coord/src/catalog/storage.rs [hunk 1]
```diff
@@ -681,25 +696,26 @@ impl<'a, S: Append> Transaction<'a, S> {
...
         cluster_name: &str,
         config: &Option<ComputeInstanceIntrospectionConfig>,
         introspection_sources: &Vec<(&'static BuiltinLog, GlobalId)>,
-    ) -> Result<ComputeInstanceId, Error> {
+    ) -> Result<(), Error> {
         let config = serde_json::to_string(config)
             .map_err(|err| Error::from(StashError::from(err.to_string())))?;
-        let id = match self.compute_instances.insert(
-            |id| ComputeInstanceKey { id: id.unwrap() },
+        match self.compute_instances.insert(
```
Comment (mjibson) [inline src/coord/src/catalog/storage.rs:706]: This should now be a `if let Err(err) = ...`

**Review by mjibson — APPROVED** [review]
Is it reasonable to add a test that does the thing that crashed?

File: src/coord/src/catalog/storage.rs (hunk 1 above)
Comment (jkosh44) [reply src/coord/src/catalog/storage.rs]: Fixed

**Comment by jkosh44** [comment]
> Is it reasonable to add a test that does the thing that crashed?

I was in the process of doing just that. Just pushed a new test to `environmentd.tests.sql`

**Review by mjibson — APPROVED** [review]

**Pull request merged**